src/main/java/com/artipie/asto/LoggingStorage.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 artipie.com
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto;
+
+import com.jcabi.log.Logger;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.logging.Level;
+import java.util.stream.Collectors;
+
+/**
+ * Storage that logs performed operations.
+ *
+ * @since 0.20.4
+ */
+public final class LoggingStorage implements Storage {
+
+    /**
+     * Logging level.
+     */
+    private final Level level;
+
+    /**
+     * Delegate storage.
+     */
+    private final Storage storage;
+
+    /**
+     * Ctor.
+     *
+     * @param storage Delegate storage.
+     */
+    public LoggingStorage(final Storage storage) {
+        this(Level.FINE, storage);
+    }
+
+    /**
+     * Ctor.
+     *
+     * @param level Logging level.
+     * @param storage Delegate storage.
+     */
+    public LoggingStorage(final Level level, final Storage storage) {
```

Comment:
@olegmoz I have a suggestion: you cold pass `Logger` to the constructor, so we can create a fake `Logger` in tests and check if `LoggingStorage` is logging what we want. WDYT?

Reply:
@paulodamaso I think we had this discussion before, I do not think that passing `Logger` instance is a good idea. Common practice is to take `Logger` instance from static context. This complicates testing (but it is still possible), but makes usage in production code much easier

Reply:
@olegmoz How could we unit test this?

Reply:
@paulodamaso first of all not everything deserves being tested. If we decide to test log output then we may use custom appender for log4j, here is example: https://stackoverflow.com/questions/3717402/how-can-i-test-with-junit-that-a-warning-was-logged-with-log4j